### PR TITLE
Typescript users.js

### PR DESCRIPTION
**Pull request by CheeseMuffin**

### CODE REVIEW

File: users.js [hunk 1]
```diff
@@ -1293,13 +1465,17 @@ class User {
 			this.chatQueueTimeout = null;
 		}
 	}
+	/**
+	 * @return {undefined}
```
Comment (Zarel) [inline users.js]: This should be unnecessary?

File: users.js [hunk 2]
```diff
@@ -1293,13 +1465,17 @@ class User {
...
+	 */
 	processChatQueue() {
 		this.chatQueueTimeout = null;
 		if (!this.chatQueue) return;
 		if (!this.chatQueue.length) {
 			this.chatQueue = null;
 			return;
 		}
+		// @ts-ignore
```
Comment (Zarel) [inline users.js]: You should not do it this way...

File: users.js [hunk 3]
```diff
@@ -413,6 +476,7 @@ class User {
 		this.inviteOnlyNextBattle = false;
 
 		// chat queue
+		/** @type {?any[]} */
```
Comment (Zarel) [inline users.js]: `@type {[string, string, Connection][]?}`

File: users.js [hunk 4]
```diff
@@ -422,15 +486,25 @@ class User {
 		this.lastMessage = ``;
 		this.lastMessageTime = 0;
 		this.lastReportTime = 0;
+		/**@type {string} */
```
Comment (Zarel) [inline users.js]: `/** @`

The rest of the code uses a space; might as well be consistent

File: users.js [hunk 5]
```diff
@@ -559,9 +655,11 @@ class User {
 	 * special permission check function is required to carry out this check
 	 * because we need to know which socket the client is connected from in
 	 * order to determine the relevant IP for checking the whitelist.
+	 * @param {Connection} connection
 	 */
 	hasConsoleAccess(connection) {
 		if (this.hasSysopAccess()) return true;
+		// @ts-ignore
```
Comment (Zarel) [inline users.js]: Wait, why this?

File: users.js (hunk 1 above)
Comment (CheeseMuffin) [reply users.js]: I think it has to have this, since it's recursive so TypeScript gets confused about the return type.

File: users.js (hunk 3 above)
Comment (CheeseMuffin) [reply users.js]: If I do it like that, TypeScript seems to be thinking later on that its of type [string | Connection][], so later on things aren't working. In processChatQueue(), with
`let [message, roomid, connection] = this.chatQueue.shift();`
it thinks connection is of type string | Connection, but idk why it thinks that / how to fix it.

File: users.js (hunk 5 above)
Comment (CheeseMuffin) [reply users.js]: w/o the ts-ignore I have to pass null parameters into `this.can`: would that be preferred (here and everywhere else with this.can)?

Comment (Zarel) [reply users.js]: Why not just let `this.can` accept default parameters?

File: users.js (hunk 1 above)
Comment (Zarel) [reply users.js]: Oh, I see.

**Comment by Zarel** [comment]
This is unfinished, but it maybe easier if I finish it up myself.

File: users.js (hunk 5 above)
Comment (CheeseMuffin) [reply users.js]: ...forgot that was a thing

**Comment by CheeseMuffin** [comment]
oh was gonna update this later, but that works too I guess.

**Comment by Zarel** [comment]
Whoa, you didn't actually TypeScript this; the `Users = any` typedef suppressed a lot of bugs here.

**Comment by Zarel** [comment]
Cheese... you really need to learn how to define default parameters in TypeScript

**Comment by Zarel** [comment]
You don't need to `@ts-ignore` every time you call `user.can` with less than 3 arguments. What you do is you define `can` like this:

```
can(permission, target = null, room = null) {
```

**Comment by CheeseMuffin** [comment]
Yeah, I’m sorry, I just totally forgot default parameters were a thing...

Are you fixing these problems or should I send another pr?

**Comment by Zarel** [comment]
I'm fixing some of the problems. You don't need to worry about it. TypeScript is currently being dumb.

**Comment by Zarel** [comment]
image

Come on, TypeScript :|

**Comment by CheeseMuffin** [comment]
yup I had that same problem idk why TypeScript can't realize the type when you explicitly define it...

**Pull request merged**